days/day-05/solutions/day05.zig

```diff
@@ -0,0 +1,95 @@
+const std = @import("std");
+const mem = std.mem;
+const io = std.io;
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+
+const MAX_FILE_SIZE = 15_000; 
+
+fn LineSegmentParser(comptime low_char: u8, comptime high_char: u8, comptime start_max: f32) type {
+    return struct {
+        fn parse(line: []const u8, from: usize, to: usize) u32 {
+            var prev: f32 = 0;
+            var min: f32 = 0;
+            var max: f32 = start_max;
+            var i: usize = from;
+        
+            while (i < to) : (i += 1) {
+                const is_not_last: bool = to - i > 1;
+
+                var delta: f32 = if (is_not_last) @ceil((max - min) / 2.0) else 0;
+
+                switch (line[i]) {
+                    low_char =>  {
+                        max -= delta;
+                        prev = min;
+                    },
+                    high_char => {
+                        min += delta;
+                        prev = max;
+                    },
+                    else => unreachable // TODO: error
+                } 
+
+            }
+
+            return @floatToInt(u32, prev);
+        }
+    };
+}
+
+pub fn main() anyerror!void {
+    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
+    defer arena.deinit();
+    const allocator = &arena.allocator;
+    
+    const in = std.io.getStdIn().reader();
+    const stdout = std.io.getStdOut().writer();
+
+    var bytes: [MAX_FILE_SIZE]u8 = undefined;
+    const length = try in.readAll(&bytes);
+
+    comptime const row_extractor = LineSegmentParser('F', 'B', 127);
+    comptime const col_extractor = LineSegmentParser('L', 'R', 7);
```

Comment:
Woow :+1:
This looks a bit like template programming, but instead it looks like the same code as everything else, instead of looking like a different language like in C++ :+1:

Reply:
Hell yeah

Reply:
Metaprogramming with uniform syntax to the main language was the main drive for zig in the beginning i believe 😄